docs/utilities/batch.md

```diff
@@ -475,6 +466,46 @@ see the details for [SQS](https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.
         }
     ```
 
+## Parallel processing
+You can choose to process batch items in parallel using the `BatchMessageHandler#processBatchInParallel()` 
+instead of `BatchMessageHandler#processBatch()`. Partial batch failure works the same way but items are processed
+in parallel rather than sequentially. 
+
+This feature is available for SQS, Kinesis and DynamoDB Streams but cannot be 
+used with SQS FIFO. In that case, items will be processed sequentially, even with the `processBatchInParallel` method.
+
+!!! warning
+    Note that parallel processing is not always better than sequential processing, 
+    and you should benchmark your code to determine the best approach for your use case. 
+
+!!! info
+    To get more threads available (more vCPUs), you need to increase the amount of memory allocated to your Lambda function.
+    While the exact vCPU allocation isn't published, from observing common patterns customers see an allocation of one vCPU per 1024 MB of memory.
```

Comment:
if you don't have multiple vCPUs (threads), it won't be parallel. For example with 512 Mb of RAM, I think you get only one... To get 2, you need ~832Mb.
I can simplify the wording...

Reply:
I don't think that's true. If you have a single vCPU and you block on IO the system scheduler will sub another process in to do work. Multithreading gives a benefit even in single core systems.

Reply:
You're right. the issue is that the `ForkJoinPool` defaults to [number of available processors](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html) threads. A thread will block on IO, so if we have 1 vCPU, we're going to run one thread for the pool, it'll block, and the function will idle until the IO comes back.

I see two options:

* `java.util.concurrent.ForkJoinPool.common.parallelism` document turning this up in the case you know you are IO-bound. If you're not doing CPU dominant stuff this'll _just work_
* Switch the batch interface up to use `CompletableFuture`s. This is a bigger piece of work, and I think would only work if the code the user wrote chained future's together, rather than doing blocking IO. So maybe not much help?

I feel like the former is probably better. Simpler and less burden on dev. Thoughts @jeromevdl ? There's probably some options with reactive streams or so too, but I have no experience there and it feels like it's getting a bit wild.

Reply:
This is what I've read at some point, could not remember where...

Also this: https://www.baeldung.com/java-when-to-use-parallel-stream#2-common-thread-pool
> It’s important to remember that this is a global setting and that it will affect all parallel streams and any other fork-join tasks that use the common pool. We strongly suggest that this parameter is not modified unless we have a very good reason for doing so.

We can probably speak about this setting, but also reinforce the message about testing because it can happen weird things: parallel processing (much) slower than sequential.

I don't think we should complexify the code, and devX with futures...

Reply:
Alternatively - how about on the builder we provide a mechanism to override the thread pool? That way if users want they can provide a customized version, and they can separate it from the common pool. This is more explicit and less "global magic" than the `-D` param. We have a nice place for it in the builder interfaces. Thoughts?

> I don't think we should complexify the code, and devX with futures...
Agreed. And if the futures do blocking stuff it doesn't improve the situation anyway.

Reply:
@jeromevdl coming back to this one - should we let the user override the thread pool and have some brief docs about this?

Reply:
In both case (`-D`) or manually in the code, it is not a best practice to do this.
See https://www.baeldung.com/java-when-to-use-parallel-stream#2-common-thread-pool & https://www.baeldung.com/java-when-to-use-parallel-stream#3-custom-thread-pool. 

Also read this: https://dzone.com/articles/be-aware-of-forkjoinpoolcommonpool. While it looks great to enable some kind of customization, it can really downgrade performance... We can maybe document the -D option so that people who know what they do use it, but we should not simplify it too much as we will receive comments that parallelism works slower than sequential...

Reply:
To be clear, I don't think we should customize the common pool, i think we should allow the user to provide a custom pool if they so desire. This opens up the ability to benefit from this PR for IO-bound tasks. The dzone link summarises this:

> if you intend to use commonPool for blocking tasks, then you need to consider some consequences. If you have more than three available CPUs, then your commonPool is automatically sized to two threads and you can very easily block execution of any other part of your system that uses the commonPool at the same time by keeping the threads in a blocked state. **As a rule of thumb, we can create our own thread-pool for blocking tasks and keep the rest of the system separated and predictable.**

If we don't want to do this, we need to make it explicit that the user should not use this change for operations that perform blocking IO in the docs. 

My thinking is 1/ Lambdas often do downstream IO-bound work that can be effectively parallelized but users will not be able to take advantage of this without a custom thread pool (or the `-D` and the flow on affects of that) and 2/ it's a pretty inoffensive escape hatch?

Reply:
Based on our discussion, we won't provide the option for custom thread pool for now. This feature can thus not be used for IO-bound work. We'll document it.

_Motivations_: 
- if we provide such a feature, some people might use it for any kind of job and expect performance to be better, while in most case it would be worse. It might gives the impression Powertools has a problem and people will open issues for this.
- By adding this feature we include a public api that would be difficult to remove in cases of miss-use (breaking change). On the other side, we might add this feature later on if people are willing to use it for IO-bound stuff and are conscious of what they do.

Reply:
Thanks for summarising @jeromevdl , and as always a good chat :)